### PR TITLE
Fix various ubsan revealed issues in the build process and regex engine.

### PR DESCRIPTION
**Pull request by demerphq**
See #19874 for related discussion. I only addressed build warnings and the regex engine.

locale.c - silence warnings about unused S_calculate_LC_ALL() under non USE_POSIX_2008_LOCALE builds
regexec.c - avoid (bogus) compiler warnings about possibly uninitialized pointer dereference and make code clearer
regcomp.h - fix non aligned pointer value stored in the regex opcode directly.
regcomp.c - deal with overflow of delta during stopmin logic related to (*ACCEPT)
regen/embed_lib.pl - deal with comments in `#else` statements better and fix handling for `#endif`

Also for the regex engine stuff I added a bunch of comments.

### CODE REVIEW

File: regcomp.c
Comment (hvds) [inline regcomp.c:6842]: necessarily

File: regcomp.c
Comment (hvds) [inline regcomp.c:6847]: Should be `AB = /(foo(*ACCEPT)|x+)whop/`

File: regcomp.c
Comment (hvds) [inline regcomp.c]: This example doesn't seem to add anything, it gives the same results as the preceding "whop" example/

File: regcomp.c
Comment (hvds) [inline regcomp.c]: I'm not sure this makes sense - the final min should be 4, I think. I also don't understand "we would have a delta of 10 ... This should result in a delta of 20".

File: regcomp.c
Comment (hvds) [inline regcomp.c]: I think this needs `... else delta = OPTIMIZE_INFTY;` (and maybe setting some is_inf* too).

File: regcomp.c
Comment (hvds) [inline regcomp.c]: Ah! This was the motivation for my original version of the fix. I note that this logic originated from the initial introduction of `is_inf_internal` by Ilya in 1998 (commit aca2d49724). At that point it was line 714!

File: regcomp.h
Comment (hvds) [inline regcomp.h:115]: In the commit message: `s/64 aligned/64 bit aligned/`; `s/that is stored/that it is stored/`; `s/which part/which is part/`

File: regcomp.h
Comment (hvds) [inline regcomp.h]: 64 bit aligned

File: regcomp.h
Comment (hvds) [inline regcomp.h]: The comment to `struct regnode_p` just above could also mention the caveats around 64-bit pointers.

File: regcomp.h
Comment (hvds) [inline regcomp.h]: This is already said in part in the preceding comment. It might be clearer to replace these extra lines with a slight augmentation of what precedes:
```
  [...] Also, there is no alignment issue,
  * because this regnode is declared or allocated as a complete unit and never
  * used in the program itself, to the compiler takes care of alignment. [...]
```

File: regcomp.h
Comment (hvds) [inline regcomp.h]: Note that this and `ARGp_SET` newly evaluate their `p` argument multiple times. Any chance of switching them to be inline functions instead?

**Review by hvds — COMMENTED** [review]
I'll leave @khwilliamson to comment on the fold_array and locale.c changes.

File: regcomp.h
Comment (tonycoz) [inline regcomp.h]: Instead of this, why not just `char ptr_bytes[sizeof(SV*)]` and memcpy() the pointer value in and out?

ARGp() would need to be (or use) an inline function or gcc brace group, but I think it would simplify things otherwise, and make fewer assumptions about pointer sizes.

The compiler can also optimize this to a simple (unaligned) access, eg.

```
#include <string.h>

typedef unsigned char U8;
typedef unsigned short U16;

struct regnode_p {
    U8	flags;
    U8  type;
    U16 next_off;
    char arg1_bytes[sizeof(void *)];
};

static inline void
ARGp_SET(struct regnode_p *n, void *p) {
    memcpy(n->arg1_bytes, &p, sizeof(p));
}

static inline void *
ARGp(const struct regnode_p *n) {
    void *p;
    memcpy(&p, n->arg1_bytes, sizeof(p));

    return p;
}

int x;

void foo(struct regnode_p *n) {
    ARGp_SET(n, &x);
}

void *bar(struct regnode_p *n) {
    return ARGp(n);
}
```
With gcc -O2, foo and bar are:
```
foo:
.LFB2:
        .cfi_startproc
        leaq    x(%rip), %rax
        movq    %rax, 4(%rdi)
        ret
        .cfi_endproc
...
bar:
.LFB3:
        .cfi_startproc
        movq    4(%rdi), %rax
        ret
        .cfi_endproc
```

As you can see, while it's undefined behaviour for us to access the unaligned data in C, the compiler has more freedom.

File: regcomp.h
Comment (demerphq) [reply regcomp.h]: Id rather not, and save the energy to eliminate regnode_p entirely. We shouldnt be putting pointers into the program like this as they wont be handled right, i believe the use case this one is put to is "special" and avoids those issues, but really we should just convert to the "Safe" api and use the data[] array.

File: regcomp.h
Comment (demerphq) [reply regcomp.h]: I dont really understand your argument. Why would using memcpy to copy 8 bytes be better than assigning two 32 bit words? 

Ideally we Just Shouldn't Have 64 bit members in our regnodes. Id rather we just eliminate the need than make our implementation better, this is just a work around for now until I have the time to address this properly and rip it out entirely. If had been paying attention when this was introduced I would have blocked it and directed Karl to the data[] array and related functionality.

File: regcomp.c
Comment (demerphq) [reply regcomp.c]: About the "4" yes, I changed this example from "foo" to "dude" and missed a number that needed to be updated as a consequence.

The part about delta is confusing because it is also off by one. But consider the min and delta /during/ the processing of this sub do NOT account for ACCEPT. The code this comment is part of is where that accounting is done. The comment is intended to describe what the initial values would be, and what we would expect as a result.

File: regcomp.c
Comment (demerphq) [reply regcomp.c]: Ah, yes, good catch.

File: regcomp.h
Comment (demerphq) [reply regcomp.h:115]: Ah. I guess I will do that in a fixup before final merge so I dont clobber this feedback with a rebase :-)

File: regcomp.h
Comment (demerphq) [reply regcomp.h]: +1

File: regcomp.h
Comment (demerphq) [reply regcomp.h]: I decided to rework that paragraph entirely as it was too dense.

File: regcomp.h
Comment (demerphq) [reply regcomp.h]: Also note that its only used in a very few places, I dont think we need to worry about it getting reused.

File: regcomp.h
Comment (demerphq) [reply regcomp.h]: I was insufficiently caffeinated when I wrote my original reply.  Roger, willco. :-)

File: regcomp.h
Comment (demerphq) [reply regcomp.h]: Ended up using the approach @tonycoz suggested, so this will use inline functions now.

File: regcomp.c
Comment (demerphq) [reply regcomp.c]: Yeah, in a future patch I think we should introduce some macros/inline functions for delta updates. I suspect there are other bugs hidden in our code that our test suite doesnt tickle. There is a conceptual issue BTW, we use "max length we can store" to represent "infinity". Which in theory could result in us turning something bounded into something unbounded.

File: regcomp.h
Comment (demerphq) [reply regcomp.h:115]: Please don't resolve this until I squash the patches.

**Comment by demerphq** [comment]
I've pushed a bunch of comment changes and fixup patches to resolve all the issues that dont need a rebase/squash.

**Comment by hvds** [comment]
> I've pushed a bunch of comment changes and fixup patches to resolve all the issues that dont need a rebase/squash.

Updates all look good to me, except I'm not sure if the inline functions should go to `inline.h` (with an appropriate prefix).

File: regcomp.h
Comment (tonycoz) [inline regcomp.h]: I used `sizeof(void *)` in my example code to keep it short(er), but this should be `sizeof(SV*)`.

Comment (demerphq) [reply regcomp.h]: Ack. Fixed.

**Comment by demerphq** [comment]
FWIW, inline.h doesn't feel right to me, these aren't general purpose and should only be available in the regex engine.

File: regcomp.h
Comment (demerphq) [reply regcomp.h:115]: Squashed now.

**Comment by demerphq** [comment]
I squashed down all the fixup patches now, and fixed the commit message @hvds pointed out.

**Comment by khwilliamson** [comment]
inline functions don't have to go into inline.h.  Only those that are needed by multiple files should go there.

File: regcomp.h
Comment (khwilliamson) [reply regcomp.h]: There's no need to be concerned with regnode_p and performance.  It is compile only, removed before execution

I had already decided that it should use a mechanism like data[], but it should use a new array that can just be wiped out at the end of compilation, since data[] is persistent.  There is no way IIUC to remove individual elements of it.

**Review by khwilliamson — APPROVED** [review]

**Review by tonycoz — APPROVED** [review]

**Comment by demerphq** [comment]
On Thu, 14 Jul 2022 at 19:27, Karl Williamson ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In regcomp.h
> <https://github.com/Perl/perl5/pull/19947#discussion_r921407585>:
>
> > +    U32 ptr_high;
> +    U32 ptr_low;
> +#else
>
> There's no need to be concerned with regnode_p and performance. It is
> compile only, removed before execution
>
> I had already decided that it should use a mechanism like data[], but it
> should use a new array that can just be wiped out at the end of
> compilation, since data[] is persistent. There is no way IIUC to remove
> individual elements of it.
>
I believe if the element is an SV and NULL then nothing would break.

Yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by demerphq** [comment]
Pushed a comment fixup, and then a squash again. Will merge after it passes CI.

**Comment by demerphq** [comment]
@hvds waiting on you to hit approve.

**Review by hvds — APPROVED** [review]

**Pull request merged**
